ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/BrokerCompatibilityCheck.java

```diff
@@ -0,0 +1,281 @@
+        .tags(metricTags);
+    final List<MetricsReporter> reporters = config.getConfiguredInstances(
+        ProducerConfig.METRIC_REPORTER_CLASSES_CONFIG,
+        MetricsReporter.class);
+    reporters.add(new JmxReporter("kafka.admin.client"));
+    final Metrics metrics = new Metrics(metricConfig, reporters, time);
+
+    final Selector selector = new Selector(
+        config.getLong(StreamsConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG),
+        metrics,
+        time,
+        "kafka-client",
+        channelBuilder,
+        logContext);
+
+    final KafkaClient kafkaClient = new NetworkClient(
+        selector,
+        metadata,
+        clientId,
+        MAX_INFLIGHT_REQUESTS, // a fixed large enough value will suffice
+        config.getLong(StreamsConfig.RECONNECT_BACKOFF_MS_CONFIG),
+        config.getLong(StreamsConfig.RECONNECT_BACKOFF_MAX_MS_CONFIG),
+        config.getInt(StreamsConfig.SEND_BUFFER_CONFIG),
+        config.getInt(StreamsConfig.RECEIVE_BUFFER_CONFIG),
+        config.getInt(StreamsConfig.REQUEST_TIMEOUT_MS_CONFIG),
+        time,
+        true,
+        new ApiVersions(),
+        logContext);
+
+    return new BrokerCompatibilityCheck(config, kafkaClient, reporters, time);
+  }
+
+  /**
+   * Check if the used brokers have version 0.10.1.x or higher.
+   *
+   * <p>Note, for <em>pre</em> 0.10.x brokers the broker version cannot be checked and the client will hang and retry
+   * until it {@link StreamsConfig#REQUEST_TIMEOUT_MS_CONFIG times out}.
+   *
+   * @throws KsqlException if brokers have version 0.10.0.x
+   */
+  void checkCompatibility() throws StreamsException {
+    final ClientRequest clientRequest = kafkaClient.newClientRequest(
+        getAnyReadyBrokerId(),
+        new ApiVersionsRequest.Builder(),
+        time.milliseconds(),
+        true);
+
+    final ClientResponse clientResponse = sendRequest(clientRequest);
+    if (!clientResponse.hasResponse()) {
+      throw new KsqlException("Received an empty response for client request when checking broker version.");
+    }
+    if (!(clientResponse.responseBody() instanceof ApiVersionsResponse)) {
+      throw new KsqlException("Inconsistent response type for API versions request. " +
+          "Expected ApiVersionsResponse but received " + clientResponse.responseBody().getClass().getName());
+    }
+
+    final ApiVersionsResponse apiVersionsResponse =  (ApiVersionsResponse) clientResponse.responseBody();
+
+    if (apiVersionsResponse.apiVersion(ApiKeys.CREATE_TOPICS.id) == null) {
+      throw new KsqlException("KSQL requires broker version 0.10.1.x or higher.");
+    }
+
+  }
+
+  /**
+   * @return the response to the request
+   * @throws TimeoutException if there was no response within {@code request.timeout.ms}
+   * @throws StreamsException any other fatal error
+   */
+  private ClientResponse sendRequest(final ClientRequest clientRequest) {
```

Comment:
Hmm.. this and the other utility methods here seem to duplicate the existing methods in `NetworkClientUtils`: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/NetworkClientUtils.java

That class already has methods for waiting for a node to be reading, doing a blocking send and receive, etc. 

Maybe we can just use that class?

Reply:
Good spot, can replace this with `NetworkClientUtils#sendAndReceive` however i don't see any value in replacing `ensureOneNodeIsReady`